### PR TITLE
Create date picker

### PR DESCRIPTION
**Pull request by PedroSoaresMoj**
Created date picker
Updated endpoints to API

### CODE REVIEW

File: server/routes/deviceController.ts [hunk 1]
```diff
@@ -13,17 +13,32 @@ export default class DeviceController {
...
 
-  async listLocations({ user, params: { deviceWearerId, deviceId } }: AuthenticatedRequest, res: Response) {
+  async listLocations(
+    { user, params: { deviceWearerId, deviceId }, query: { startDate = '', endDate = '' } }: AuthenticatedRequest,
+    res: Response,
+  ) {
     const validatedWearerId = deviceWearerId.toString()
     const validatedDeviceId = deviceId.toString()
+
+    const validStartDate = startDate ? startDate.toString() : '' // AddYears(-7, new Date()).toISOString().slice(0, 16)
```
Comment (JazJax) [inline server/routes/deviceController.ts]: We should remove the commented code

File: server/routes/deviceRouter.ts [hunk 2]
```diff
@@ -3,6 +3,7 @@ import { Services } from '../services'
 import asyncMiddleware from '../middleware/asyncMiddleware'
 import DeviceController from './deviceController'
 import { AuthenticatedRequest } from '../authentication/auth'
+// import { post } from 'superagent'
```
Comment (JazJax) [inline server/routes/deviceRouter.ts]: Remove commented code

File: server/data_models/device.ts [hunk 3]
```diff
@@ -8,9 +8,13 @@ type DeviceType =
 type DeviceStatus = 'Not fitted' | 'Fitted' | 'Removed'
 
 type Device = {
+  id: number
```
Comment (ben-cart3r) [inline server/data_models/device.ts:11]: Why do we have an additional field that isn't in the data dictionary schema?

File: server/routes/deviceController.ts [hunk 4]
```diff
@@ -13,17 +13,32 @@ export default class DeviceController {
...
+        validatedDeviceId,
+        validStartDate,
+        validEndDate,
+      )
       this.renderDeviceLocationListView(res, {
         isError: false,
         locations,
         backLink: `/device-wearers/id/${validatedWearerId}`,
         device,
+        searchUrl: `/device/${deviceWearerId}/${deviceId}`,
```
Comment (ben-cart3r) [inline server/routes/deviceController.ts]: I don't think this is required, you can leave the form action empty

File: server/viewModels/device.ts [hunk 5]
```diff
@@ -4,7 +4,14 @@ import Location from '../data_models/location'
 
 type DeviceDetailViewModel =
   | BaseErrorModel
-  | (BaseErrorModel & { device: Device; backLink: string })
-  | (BaseSuccessModel & { device: Device; locations: Array<Location>; backLink: string })
+  | (BaseErrorModel & { device: Device; backLink: string; searchUrl: string })
```
Comment (JazJax) [inline server/viewModels/device.ts]: Have you confirmed the error model loads correctly given it doesn't have the date parameters passed at all?

**Review by JazJax — APPROVED** [review]
Couple of comments/cleanup, but lgtm, happy for merge.

File: server/services/deviceService.ts [hunk 6]
```diff
@@ -21,10 +21,10 @@ export default class DeviceService {
 
   async findDeviceById(token: string, deviceId: string): Promise<Device> {
     try {
-      const response = await this.restClient(token).get<ApiResponse<'device', Device>>({
+      const response = await this.restClient(token).get<ApiResponse<'devices', Device[]>>({
```
Comment (ben-cart3r) [inline server/services/deviceService.ts:24]: This is more a question for the API, but why is it returning a list for a request that should return 1 result?

File: server/services/locationService.test.ts [hunk 7]
```diff
@@ -8,13 +8,15 @@ describe('location service', () => {
...
+      latitude: 20.0,
+      longitude: 20.0,
       locationTime: '2023:12:12T12:34:00',
     },
     {
-      latitude: '20.0',
-      longitude: '20.0',
+      id: 2,
+      latitude: 20.0,
+      longitude: 20.0,
```
Comment (ben-cart3r) [inline server/services/locationService.test.ts:19]: Remove fake data or implement a dummy service for testing that can be injected to the controllers.

File: server/services/locationService.test.ts [hunk 8]
```diff
@@ -27,17 +29,19 @@ describe('location service', () => {
...
-    it('should call API to get all locations for device Id', async () => {
+    const endpointUrl = `/locations/v1/device-id/${deviceId}`
+    const startDate = ''
+    const endDate = ''
+    it('should call API to get all locations for device Id when start and endDate are empty', async () => {
       const expected = dummyData
       let called = 0
 
       fakeDataPlatformApi.get(endpointUrl).reply(200, () => {
         called += 1
```
Comment (ben-cart3r) [inline server/services/locationService.test.ts]: I thought we were removing this?

File: server/viewModels/device.ts (hunk 5 above)
Comment (ben-cart3r) [reply server/viewModels/device.ts]: Cypress tests!!!

File: server/views/pages/device/detail.njk [hunk 9]
```diff
@@ -95,7 +94,13 @@
     ]
   })
   }}
-
+  <form action= {{ searchUrl }}>
```
Comment (ben-cart3r) [inline server/views/pages/device/detail.njk]: As above, the action attribute shouldn't be needed

File: server/data_models/device.ts (hunk 3 above)
Comment (PedroSoaresMoj) [reply server/data_models/device.ts:11]: to reflect the same fields that exist in the API

Comment (ben-cart3r) [reply server/data_models/device.ts:11]: I guess my point is why are they in the API?

File: server/services/deviceService.ts (hunk 6 above)
Comment (PedroSoaresMoj) [reply server/services/deviceService.ts:24]: we are always returning lists from the API, they can be empty or have 1 or more elements.

**Pull request merged**